### PR TITLE
Fix MLX backend divergence from PyTorch on stitching, CPM RevIN refine, residual activation, and NaN handling

### PR DESCRIPTION
**Pull request by matdou**
I found four places where the MLX backend for TimesFM3 computes something different from PyTorch. I copied a small PyTorch model's weights into an equivalent MLX model (safetensors round-trip) and compared `decode()` output.

1. `decode()` always used the stitching-based horizon formula and never checked `cfg.use_stitching` (even though `mlx/configs.py` parse the flag). Fixed by branching on `cfg.use_stitching`, mirroring the non-stitching padding/extraction code from `torch/model.py`.

2. The CPM RevIN refine step ran unconditionally. `TimesFM3MlxConfig` had no field to turn it off, unlike torch's `use_iterative_cpm_revin`. Added the field.

3. `ResidualBlock` hardcoded ReLU and ignored the configured `activation`, `prenorm`, and `identity_skip`. So a checkpoint using `swish` or RMSNorm prenorm would run the wrong function.

4. Inputs weren't sanitized for NaN before computing running stats. Added the same `nan_to_num` + clip step torch runs before RevIN.

Added `mlx/model_parity_test.py`: builds tiny matched models in both backends, transplants weights, and diffs `decode()` output. It covers configs both backends and  reproduces the four bugs; before the fix these failed with real numeric gaps (3.68, 0.053, 0.997, NaN-vs-finite), after the fix they pass.

Test plan

- [x] `pytest src/timesfm3/mlx/model_parity_test.py` -- 6 passed
- [x] `pytest src/timesfm3/` -- 64 passed, 14 skipped (checkpoint-dependent tests, unaffected by this change)
- [x] Verified no regression on adjacent flags not touched by this fix (`use_variate_attention=False`, `use_linear_detrending=False`) by the same weight-transplant method

Known gap not covered here: the same class of bug exists in a few other transformer config fields (`ff_activation`, `v_norm`, `causal_attention`, `use_rope_var`, `use_memory_efficient_attention`). MLX still hardcodes its behavior for those. I can follow up with a second PR if that's useful.

@rachittshah

### CODE REVIEW

**Comment by matdou** [comment]
Went ahead and covered the "known gap" from the description in this same PR instead of a separate one — added a commit that fixes `ff_activation`, `v_norm`, `causal_attention`, `use_rope_var`, and `use_memory_efficient_attention`. Same root cause each time: `mlx/transformer.py` hardcoded torch's default for these instead of reading the config.

Found one more thing while wiring `v_norm` in: the existing single-variate attention shortcut (`n == 1` skips straight to the value projection, since softmax over one key is always 1.0) was fine for Q/K/RoPE/softmax but silently wrong for value normalization, since `v_norm` changes the value itself. Fixed that too.

`use_bias` and `qk_norm="none"` are different from the rest — they change which weights exist, so MLX already fails to load with a clear key-mismatch error instead of silently diverging. Left those alone.

Verified the same way as the first commit: weight-transplant + diff against torch, one test per flag, plus the full suite (69 passed, 14 skipped, all pre-existing skips).

**Comment by rajatsen91** [comment]
Thanks for the important PR, @matdou! I have a few suggestions detailed below.

### 1. Longer Horizon Coverage for CPM RevIN Refine
In `TimesFM3KnownDivergenceTest` (`test_use_iterative_cpm_revin_false_is_ignored_by_mlx`), with `_OUTPUT_PATCH_LEN = 16` and `_INPUT_PATCH_LEN = 8` (`rolls = 2`), `horizon = 24` only exercises the first boundary patch into the CPM region.

Could we expand or add a test with a longer horizon (such as `horizon = 48` or `64`)? A longer horizon forces `cpm_iterative_revin_refine` to execute across multiple block offsets (`block_offset = (block_offset + 1) % rolls`) and accumulate `anchor` predictions across consecutive rolled steps, ensuring multi-step iterative refinement is tested against PyTorch:

```python
  def test_use_iterative_cpm_revin_long_horizon(self):
    # Tests that multi-step iterative refinement across multiple rolled blocks
    # matches torch when enabled (and also when disabled).
    for use_cpm in (True, False):
      torch_model, mlx_model = _build_pair(use_iterative_cpm_revin=use_cpm)
      ctx = np.random.RandomState(0).randn(1, 1, 128).astype(np.float32)
      torch_out, mlx_out = _decode_both(torch_model, mlx_model, ctx, horizon=48)
      np.testing.assert_allclose(torch_out, mlx_out, atol=1e-4)
```

---

### 2. CPM RevIN Refine With & Without Covariates
`cpm_iterative_revin_refine` handles tensor shapes `(b, v, rolls, patch_len)` where `v` combines targets and covariates (`num_pred = num_target + num_past_only`).

Could we add a test verifying that `use_iterative_cpm_revin=True` (and `False`) works identically with `past_only_covariates` and `past_future_covariates`?

```python
  def test_use_iterative_cpm_revin_with_covariates(self):
    rng = np.random.RandomState(42)
    target = rng.randn(1, 2, 128).astype(np.float32)
    po = rng.randn(1, 1, 128).astype(np.float32)
    pf = rng.randn(1, 1, 128 + 32).astype(np.float32)
    
    for use_cpm in (True, False):
      torch_model, mlx_model = _build_pair(use_iterative_cpm_revin=use_cpm)
      torch_out, mlx_out = _decode_both(
        torch_model,
        mlx_model,
        target,
        horizon=32,
        past_only_covariates=po,
        past_future_covariates=pf,
      )
      np.testing.assert_allclose(torch_out, mlx_out, atol=1e-4)
```

---

### 3. Snapshot Fallback for `_weights_cached()` (in `timesfm3_forecaster_test.py`)
In `timesfm3_forecaster_test.py`, `try_to_load_from_cache(_CHECKPOINT, "model.safetensors")` checks `refs/main`. If `main` is updated on the Hub (for example, when a README is edited), `refs/main` advances to a new commit that may not yet have the `model.safetensors` symlink created, causing `try_to_load_from_cache` to return `None` even when `model.safetensors` (~4 GB) is already present in an earlier snapshot.

We can add a simple fallback to inspect the snapshot directory:

```python
def _weights_cached() -> bool:
  try:
    from huggingface_hub import try_to_load_from_cache

    path = try_to_load_from_cache(_CHECKPOINT, "model.safetensors")
    if isinstance(path, str) and os.path.exists(path):
      return True
    # Fallback: check if model.safetensors exists in any cached snapshot
    repo_dir = os.path.expanduser(
      f"~/.cache/huggingface/hub/models--{_CHECKPOINT.replace('/', '--')}/snapshots"
    )
    if os.path.isdir(repo_dir):
      return any(
        os.path.exists(os.path.join(repo_dir, snap, "model.safetensors"))
        for snap in os.listdir(repo_dir)
      )
    return False
  except Exception:
    return False
```

---

### 4. (Minor Polish) Add `"silu"` Alias in `src/timesfm3/mlx/dense.py`
In `dense.py`, `_ACTIVATIONS` defines `"swish"`, `"relu"`, and `"none"`. In `timesfm3.torch.util.get_activation_fn`, `"silu"` is also supported as an alias to SiLU/Swish:
```python
_ACTIVATIONS = {
  "relu": nn.relu,
  "swish": nn.silu,
  "silu": nn.silu,
  "none": lambda x: x,
}
```

**Comment by matdou** [comment]
Thank you for your insights ! I wrote 1 and 2. Both matched to ~1e-6. I fixed 4.

3. the fix is in `timesfm3_forecaster_test.py`, which this PR doesn't touch. split it into #510.
